src/toil/lib/io.py

```diff
@@ -63,6 +73,11 @@ def robust_rmtree(path: Union[str, bytes]) -> None:
         except FileNotFoundError:
             # File went away
             return
+        except OSError as exc:
+            if exc.errno == 16:
+                # 'Device or resource busy'
+                return
```

Comment:
Hmmm... would it be better for this function to be retrying for a short period of time on OSError 16 rather than immediately returning?

Reply:
> > Based on some additional stack traces, I think I found something else:
> > https://github.com/DataBiosphere/toil/blob/master/src/toil/jobStores/fileJobStore.py#L260-L265 https://stackoverflow.com/questions/41362016/rename-atomicity-and-nfs
> > This seems like a problem. Maybe those lines can be wrapped in a `global_mutex`?
> 
> That sounds like a good idea to me. Might be worth making a separate PR just to keep things simple. Thanks again.

Yeah, that's a bigger change so I was gonna keep it separate.

Reply:
I just blindly copied the above except clause. You're probably right that retrying would be better.

Reply:
Okay, I removed the `except OSError` clauses from `robust_rmtree` and added `@retry(errors=[OSError])`, but now my log file looks like this:

```
[2023-04-20T13:41:59+0000] [MainThread] [W] [toil.lib.retry] Error in <function robust_rmtree at 0x00007fe76085ce80>: [Errno 16] Device or resource busy: b'/home/fennickjr/coorddir/57d4a87a1044516e91e442d91dae76ad/deferred/.panfs.b0810ac.1682012487623425000'. Retrying after 1 s...
[2023-04-20T13:42:00+0000] [MainThread] [W] [toil.lib.retry] Error in <function robust_rmtree at 0x00007fe76085ce80>: [Errno 16] Device or resource busy: b'/home/fennickjr/coorddir/57d4a87a1044516e91e442d91dae76ad/deferred/.panfs.b0810ac.1682012487623425000'. Retrying after 1 s...
[2023-04-20T13:42:01+0000] [MainThread] [W] [toil.lib.retry] Error in <function robust_rmtree at 0x00007fe76085ce80>: [Errno 16] Device or resource busy: b'/home/fennickjr/coorddir/57d4a87a1044516e91e442d91dae76ad/deferred/.panfs.b0810ac.1682012487623425000'. Retrying after 1 s...
[2023-04-20T13:42:02+0000] [MainThread] [W] [toil.lib.retry] Error in <function robust_rmtree at 0x00007fe76085ce80>: [Errno 16] Device or resource busy: b'/home/fennickjr/coorddir/57d4a87a1044516e91e442d91dae76ad/deferred/.panfs.b0810ac.1682012487623425000'. Retrying after 2 s...
[2023-04-20T13:42:04+0000] [MainThread] [W] [toil.lib.retry] Error in <function robust_rmtree at 0x00007fe76085ce80>: [Errno 16] Device or resource busy: b'/home/fennickjr/coorddir/57d4a87a1044516e91e442d91dae76ad/deferred/.panfs.b0810ac.1682012487623425000'. Retrying after 4 s...
[2023-04-20T13:42:08+0000] [MainThread] [W] [toil.lib.retry] Error in <function robust_rmtree at 0x00007fe76085ce80>: [Errno 16] Device or resource busy: b'/home/fennickjr/coorddir/57d4a87a1044516e91e442d91dae76ad/deferred/.panfs.b0810ac.1682012487623425000'. Retrying after 8 s...
[2023-04-20T13:42:16+0000] [MainThread] [W] [toil.lib.retry] Error in <function robust_rmtree at 0x00007fe76085ce80>: [Errno 16] Device or resource busy: b'/home/fennickjr/coorddir/57d4a87a1044516e91e442d91dae76ad/deferred/.panfs.b0810ac.1682012487623425000'. Retrying after 16 s...
...
[2023-04-20T13:42:32+0000] [MainThread] [W] [toil.lib.retry] Error in <function robust_rmtree at 0x00007fe76085ce80>: [Errno 16] Device or resource busy: b'/home/fennickjr/coorddir/57d4a87a1044516e91e442d91dae76ad/deferred/.panfs.b0810ac.1682012487623425000'. Retrying after 1 s...
[2023-04-20T13:42:33+0000] [MainThread] [W] [toil.lib.retry] Error in <function robust_rmtree at 0x00007fe76085ce80>: [Errno 16] Device or resource busy: b'/home/fennickjr/coorddir/57d4a87a1044516e91e442d91dae76ad/deferred/.panfs.b0810ac.1682012487623425000'. Retrying after 1 s...
[2023-04-20T13:42:34+0000] [MainThread] [W] [toil.lib.retry] Error in <function robust_rmtree at 0x00007fe76085ce80>: [Errno 16] Device or resource busy: b'/home/fennickjr/coorddir/57d4a87a1044516e91e442d91dae76ad/deferred/.panfs.b0810ac.1682012487623425000'. Retrying after 1 s...
[2023-04-20T13:42:35+0000] [MainThread] [W] [toil.lib.retry] Error in <function robust_rmtree at 0x00007fe76085ce80>: [Errno 16] Device or resource busy: b'/home/fennickjr/coorddir/57d4a87a1044516e91e442d91dae76ad/deferred/.panfs.b0810ac.1682012487623425000'. Retrying after 2 s...
[2023-04-20T13:42:37+0000] [MainThread] [W] [toil.lib.retry] Error in <function robust_rmtree at 0x00007fe76085ce80>: [Errno 16] Device or resource busy: b'/home/fennickjr/coorddir/57d4a87a1044516e91e442d91dae76ad/deferred/.panfs.b0810ac.1682012487623425000'. Retrying after 4 s...
[2023-04-20T13:42:41+0000] [MainThread] [W] [toil.lib.retry] Error in <function robust_rmtree at 0x00007fe76085ce80>: [Errno 16] Device or resource busy: b'/home/fennickjr/coorddir/57d4a87a1044516e91e442d91dae76ad/deferred/.panfs.b0810ac.1682012487623425000'. Retrying after 8 s...
[2023-04-20T13:42:49+0000] [MainThread] [W] [toil.lib.retry] Error in <function robust_rmtree at 0x00007fe76085ce80>: [Errno 16] Device or resource busy: b'/home/fennickjr/coorddir/57d4a87a1044516e91e442d91dae76ad/deferred/.panfs.b0810ac.1682012487623425000'. Retrying after 16 s...
[2023-04-20T13:45:43+0000] [MainThread] [E] [toil.deferred] [Errno 16] Device or resource busy: b'/home/fennickjr/coorddir/57d4a87a1044516e91e442d91dae76ad/deferred/.panfs.b0810ac.1682012487623425000'
Traceback (most recent call last):
  File "/home/fennickjr/mambaforge-pypy3/envs/wic/lib/pypy3.9/site-packages/toil/deferred.py", line 214, in cleanupWorker
    robust_rmtree(os.path.join(stateDirBase, cls.STATE_DIR_STEM))
  File "/home/fennickjr/mambaforge-pypy3/envs/wic/lib/pypy3.9/site-packages/toil/lib/retry.py", line 292, in call
    return func(*args, **kwargs)
  File "/home/fennickjr/mambaforge-pypy3/envs/wic/lib/pypy3.9/site-packages/toil/lib/io.py", line 57, in robust_rmtree
    robust_rmtree(child_path)
  File "/home/fennickjr/mambaforge-pypy3/envs/wic/lib/pypy3.9/site-packages/toil/lib/retry.py", line 292, in call
    return func(*args, **kwargs)
  File "/home/fennickjr/mambaforge-pypy3/envs/wic/lib/pypy3.9/site-packages/toil/lib/io.py", line 75, in robust_rmtree
    os.unlink(path)
OSError: [Errno 16] Device or resource busy: b'/home/fennickjr/coorddir/57d4a87a1044516e91e442d91dae76ad/deferred/.panfs.b0810ac.1682012487623425000'
```

That retry block is repeating many times, and after about 5 minutes it throws the stack trace anyway. Is this what we want? And it doesn't just do this once. This whole process repeats for various other lock files, so adding in this retry is killing throughput.

My cluster administrator is not yet convinced there's anything wrong with our cluster and/or with the panasas file system, but IDK.

Reply:
I think there's a way to write an `ErrorCondition` that matches only some errors of a particular type. I'm not sure if it would work here.